keps/sig-apps/0010-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,157 @@
+
+	// Rolling update config params. Present only if DeploymentStrategyType =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// +optional
+	RollingUpdate *RollingUpdateDeployment
+
++	// InPlace update config params. Present only if DeploymentStrategyType =
++	// InPlaceUpdate.
++	// +optional
++	InPlaceUpdate *InPlaceUpdateDeployment
+}
+
+const(
+	// Kill all existing pods before creating new ones.
+	RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+
+        // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+        RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+    
+        // Update containers inside the pods without tearing down the pods
++	InPlaceUpdateDeploymentStrategyType DeploymentStrategyType = "InPlaceUpdate"
+)
+
+// Spec to control the desired behavior of in-place update
+type InPlaceUpdateDeployment struct {
+	// The number of pods to be updated in each batch.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
+	// Absolute number is calculated from percentage by rounding down.
+	BatchSize intstr.IntOrString
+}
+```
+
+User specifies the new update strategy type in the spec and issue the request
+
+### Deployment Controller
+
+#### Update
+Below are two alternative approaches for in-place update in simplified words. 
+Wait until expected number of pods created, when deployment controller receives the in-place update request
+
+##### Approach 1
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+    
+    // Pause the old replica set. This is required for this scenario: if we reduce oldReplicaSet immediately, there's a chance
+    // that oldReplicaSet may kill the pods prematurely.
+    pause(oldReplicaSet)
+    
+    // Reduce oldReplicaSet by the BatchSize
+    update(oldReplicaSet, replicas - BatchSize)
+    
+    // update pod controllerRef to the new ReplicaSet
+    for pod range(podsToUpdate)
+        update(pod, newReplicaSet-controllerRef)
+
+    // Increase newReplicaSet by BatchSize
+    update(newReplicaSet, replicas + BatchSize)
+
+```
+
+##### Approach 2
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+
+    // swap oldReplicaSet and newReplicaSet information such as spec, revision, annotation etc.
```

Comment:
This swap is not atomic and will not play well with the deployment controller and the mechanism it uses for creating new replicasets, though I know @janetkuo has a proposal to change that.

Reply:
Could you elaborate what the issue is ? Since the new ReplicaSet is created with 0 replicas, IIUC, this behaves like a no-op, even though the swap is not atomic. the flow is like below:
1. oldCopy-info = oldReplicaSet-info // write the copy into etcd,  in case the controller failed, 
2. set(oldReplicaSet, newReplicaSet-info)
3. set(newReplicaSet, oldCopy-info)